contracts/cw20-ics20/src/ibc.rs

```diff
@@ -243,12 +220,41 @@ fn do_ibc_packet_receive(deps: DepsMut, packet: &IbcPacket) -> Result<Ics20Packe
             let mut cur = orig.ok_or(ContractError::InsufficientFunds {})?;
             cur.outstanding = cur
                 .outstanding
-                .checked_sub(amount)
+                .checked_sub(msg.amount)
                 .or(Err(ContractError::InsufficientFunds {}))?;
             Ok(cur)
         },
     )?;
-    Ok(msg)
+
+    let to_send = Amount::from_parts(denom.to_string(), msg.amount);
+    let gas_limit = check_gas_limit(deps.as_ref(), &to_send)?;
+    let send = send_amount(to_send, msg.receiver.clone(), gas_limit);
+
+    let res = IbcReceiveResponse::new()
+        .set_ack(ack_success())
+        .add_submessage(send)
+        .add_attribute("action", "receive")
+        .add_attribute("sender", msg.sender)
+        .add_attribute("receiver", msg.receiver)
+        .add_attribute("denom", denom)
+        .add_attribute("amount", msg.amount)
+        .add_attribute("success", "true");
+
+    Ok(res)
+}
+
+fn check_gas_limit(deps: Deps, amount: &Amount) -> Result<Option<u64>, ContractError> {
+    match amount {
+        Amount::Cw20(coin) => {
+            // if cw20 token, use the registered gas limit, or error if not whitelisted
+            let addr = deps.api.addr_validate(&coin.address)?;
+            Ok(ALLOW_LIST
+                .may_load(deps.storage, &addr)?
+                .ok_or(ContractError::NotOnAllowList)?
+                .gas_limit)
+        }
+        _ => Ok(None),
```

Comment:
What about these? Just curious.

Reply:
Native token transfers go to bank. No need to enforce limits.
That logic is trusted cosmos sdk code.

This is defines about calling into a cw20 token contract that might do ANYTHING, especially burning all gas in this transaction which would cause trouble with the relayer